sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/UnconditionalJumpStatementBase.cs

```diff
@@ -0,0 +1,149 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Linq;
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules
+{
+    public abstract class UnconditionalJumpStatementBase<TStatementSyntax, TLanguageKindEnum> : SonarDiagnosticAnalyzer
+        where TStatementSyntax : SyntaxNode
+        where TLanguageKindEnum : struct
+    {
+        protected const string DiagnosticId = "S1751";
+        protected const string MessageFormat = "Remove this '{0}' statement or make it conditional.";
+
+        protected abstract DiagnosticDescriptor Rule { get; }
+        protected abstract GeneratedCodeRecognizer GeneratedCodeRecognizer { get; }
+
+        protected abstract LoopWalkerBase<TStatementSyntax, TLanguageKindEnum> GetWalker(SyntaxNodeAnalysisContext context);
+        protected abstract ISet<TLanguageKindEnum> LoopStatements { get; }
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(GeneratedCodeRecognizer,
+                c =>
+                {
+                    var walker = GetWalker(c);
+                    walker.Visit();
+                    foreach (var node in walker.GetRuleViolations())
+                    {
+                        c.ReportDiagnostic(Diagnostic.Create(Rule, node.GetLocation(), GetKeywordText(node)));
+                    }
+                },
+                LoopStatements.ToArray());
+        }
+
+        protected abstract string GetKeywordText(TStatementSyntax statement);
+    }
+
+    public abstract class LoopWalkerBase<TStatementSyntax, TLanguageKindEnum>
+        where TStatementSyntax : SyntaxNode
+        where TLanguageKindEnum : struct
+    {
+        protected readonly SyntaxNode rootExpression;
+
+        protected List<TStatementSyntax> ConditionalContinues { get; } = new List<TStatementSyntax>();
+        protected List<TStatementSyntax> ConditionalTerminates { get; } = new List<TStatementSyntax>();
+
+        protected List<TStatementSyntax> UnconditionalContinues { get; } = new List<TStatementSyntax>();
+        protected List<TStatementSyntax> UnconditionalTerminates { get; } = new List<TStatementSyntax>();
+
+        protected abstract ISet<TLanguageKindEnum> ConditionalStatements { get; }
+        protected abstract ISet<TLanguageKindEnum> StatementsThatCanThrow { get; }
+        protected abstract ISet<TLanguageKindEnum> LambdaSyntaxes { get; }
+
+        private ISet<TLanguageKindEnum> lambdaOrLoopStatements;
+        private ISet<TLanguageKindEnum> loopStatements;

```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Make 'loopStatements' 'readonly'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS2933)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove the 'loopStatements' field and declare it as a local variable in the relevant methods. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1450)